src/object/cli_obj.c

```diff
@@ -6646,6 +6649,338 @@ shard_punch_prep(struct shard_auxi_args *shard_auxi, struct dc_object *obj,
+		tgt_ids[node_nr] = NULL;
+	}
+
+	D_ALLOC(mbs, sizeof(*mbs) + sizeof(*ddt) * count);
+	if (mbs == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	/*
+	 * For object collective punch, we always commit related DTX synchronously. Even if we lost
+	 * some redundancy groups when DTX resync, we still continue to punch the remaining shards.
+	 * So set dm_grp_cnt as 1 to bypass redundancy group check.
+	 */
+	mbs->dm_grp_cnt = 1;
+	mbs->dm_tgt_cnt = count;
+	mbs->dm_data_size = sizeof(*ddt) * count;
+	mbs->dm_flags = DMF_CONTAIN_LEADER;
+
+	/* ddt[0] will be the lead target. */
+	ddt = &mbs->dm_tgts[0];
+	for (i = 0, j = 0; i < *length; i++) {
+		oct = &octs[i];
+		for (m = 0, n = 0; n < oct->oct_shard_nr; m++) {
+			if (isset(oct->oct_bitmap, m)) {
+				ddt[j++].ddt_id = tgt_ids[i][m];
+				n++;
+			}
+		}
+	}
+
+	D_ASSERTF(j == count, "Invalid targets count %d vs %d\n", count, j);
+
+out:
+	if (tgt_ids != NULL) {
+		for (i = 0; i < node_nr + 1; i++)
+			D_FREE(tgt_ids[i]);
+		D_FREE(tgt_ids);
+	}
+
+	if (shard != NULL)
+		obj_shard_close(shard);
+
+	if (rc < 0) {
+		if (octs != NULL)
+			obj_coll_punch_cleanup(mbs, octs, node_nr + 1);
+		*length = 0;
+	} else {
+		*p_octs = octs;
+		*p_mbs = mbs;
+	}
+
+	return rc;
+}
+
+struct obj_coll_punch_cb_args {
+	crt_rpc_t	*ocpca_rpc;
+	uint32_t	*ocpca_ver;
+};
+
+static int
+obj_coll_punch_cb(tse_task_t *task, void *data)
+{
+	struct obj_coll_punch_cb_args	*cb_args = data;
+	crt_rpc_t			*rpc = cb_args->ocpca_rpc;
+	struct obj_coll_punch_in	*ocpi = crt_req_get(rpc);
+
+	if (task->dt_result == 0) {
+		task->dt_result = obj_reply_get_status(rpc);
+		*cb_args->ocpca_ver = obj_reply_map_version_get(rpc);
+	}
+
+	DL_CDEBUG(task->dt_result < 0, DLOG_ERR, DB_IO, task->dt_result,
+		  "DAOS_OBJ_RPC_COLL_PUNCH RPC %p for "DF_UOID" on %d ranks with DTX "
+		  DF_DTI" for task %p, map_ver %u/%u, flags %lx/%x\n",
+		  rpc, DP_UOID(ocpi->ocpi_oid), (int)ocpi->ocpi_tgts.ca_count,
+		  DP_DTI(&ocpi->ocpi_odm.odm_xid), task, ocpi->ocpi_map_ver, *cb_args->ocpca_ver,
+		  (unsigned long)ocpi->ocpi_api_flags, ocpi->ocpi_flags);
+
+	obj_coll_punch_cleanup(ocpi->ocpi_odm.odm_mbs, ocpi->ocpi_tgts.ca_arrays,
+			       ocpi->ocpi_tgts.ca_count);
+	crt_req_decref(rpc);
+
+	return task->dt_result;
+}
+
+static int
+dc_obj_coll_punch(tse_task_t *task, struct dc_object *obj, struct dtx_epoch *epoch, uint32_t map_ver,
+		  daos_obj_punch_t *args, struct obj_auxi_args *auxi)
+{
+	struct dc_pool			*pool = obj->cob_pool;
+	struct obj_coll_target		*octs = NULL;
+	struct dtx_memberships		*mbs = NULL;
+	crt_rpc_t			*req = NULL;
+	struct obj_coll_punch_in	*ocpi = NULL;
+	struct obj_coll_punch_cb_args	 cb_args = { 0 };
+	crt_endpoint_t			 tgt_ep = { 0 };
+	uint32_t			 length = 0;
+	int				 rc;
+	int				 i;
+
+	rc = obj_coll_punch_prep(obj, map_ver, &mbs, &octs, &length);
+	if (rc < 0)
+		goto out;
+
+	tgt_ep.ep_grp = pool->dp_sys->sy_group;
+	tgt_ep.ep_rank = octs[0].oct_rank;
+
+	rc = octs[0].oct_bitmap_sz << 3;
```

Comment:
D_ASSERT(octs[0].oct_shard_nr >= 2); ? (or the "octs[0].oct_shard_nr--" maybe wrong).

Reply:
The octs[0] is the leader. When elect leader inside obj_coll_punch_prep(), we only choose the one with "oct_shard_nr > 1" at L6793.

Reply:
ok, i saw that